### ISSUE TITLE
Improve support for switching between dev and production composer builds

### ISSUE THREAD

**Issue by deviantintegral** [Lullabot/drainpipe#169]
In my mind, one of the biggest benefit of `task` is that it allows us to express:

1. How to determine if source files have changed, and therefore a command needs to be re-run.
2. Dependencies that can be run in parallel, and those that need to be run in serial.

As is, the current `drupal:update` task has no dependency on composer. Ideally, if composer.json or composer.lock has changed, this task will run composer install automatically: https://github.com/Lullabot/drainpipe/blob/main/tasks/drupal.yml#L56-L58

We've replaced a few tasks in our site-specific Taskfile that I've included below. As well, by consolidating to a single command for `composer install` for both prod and dev installs, it allows parent tasks to depend on a single task and set the environment variable if needed.

What do others think about an approach like this? If it looks good, I can file a PR (but probably after the holidays).

cc @hawkeyetwolf

```yaml
  # Override Drainpipe's drupal:composer and drupal:update tasks until they
  # are changed upstream.
  #
  # The drupal:update command has no dependency on composer jobs. There's two
  # cases where we need composer to be run:
  #  1. Where composer hasn't been run at all.
  #  2. Where an environment is changing the "no-dev" flag for composer install.
  #
  # Since composer already has an environment variable to use, we use that
  # instead of having separate jobs. This allows drupal:update to depend on a
  # single composer task definition.
  override:drupal:composer:
    desc: Install composer dependencies
    summary: |
      This command is typically used when a developer updates their local after
      pulling new changes from source control. Generally when bootstrapping a
      fresh clone of the site from git, you'll need to run `composer install`
      anyway to get task and drainpipe.
    cmds:
      - |
        # We use the COMPOSER_NO_DEV environment variable to control production
        # builds.
        # https://getcomposer.org/doc/03-cli.md#composer-no-dev
        # We set it in case task is using the "alternate" syntax, including
        # when we call task via `ddev task`.
        export COMPOSER_NO_DEV={{.COMPOSER_NO_DEV}}
        composer install --optimize-autoloader
    sources:
      - composer.json
      - composer.lock
    generates:
      - ./vendor/composer/installed.json
      - ./vendor/autoload.php
    status:
      - >
        {{if eq .COMPOSER_NO_DEV "1"}}
        test -f ./vendor/composer/installed.json && grep -q '"dev": false' ./vendor/composer/installed.json
        {{else}}
        test -f ./vendor/composer/installed.json && grep -q '"dev": true' ./vendor/composer/installed.json
        {{end}}

  override:drupal:update:
    desc: Always run composer before updating.
    deps: [override:drupal:composer, drupal:update]

  build:
    desc: "Builds the project for production"
    cmds:
      - composer validate
      - task: override:drupal:composer
        vars:
          COMPOSER_NO_DEV: 1

  build:dev:
    desc: "Builds the project for development use"
    deps: [override:drupal:composer, assets:dev]
    cmds:
      # Task always calls dependencies in parallel. Since we need composer to
      # finish first, we run this in serial.
      # https://taskfile.dev/usage/#task-dependencies
      - task: override:drupal:update
      - ./vendor/bin/drush {{ .site }} -y en project_devel
```

**Comment by justafish** [comment]
Just having a play locally it's saving around 3 seconds for me vs calling composer install the regular way, but I think this would need to stay at the project level as there's scenarios where you don't want composer to be run e.g.
- When `drupal:update` is used in a remote environment
- Projects using composer scripts that still do things on install/update regardless of the state of dependencies
- Trying to debug problems in upstream dependencies

**Comment by deviantintegral** [comment]
> Just having a play locally it's saving around 3 seconds for me vs calling composer install

Is that on a regular Drupal core install, or a full client site? When we initially added this, I know it was saving at least 5-10 seconds for me, and more for developers with older workstations.

> When drupal:update is used in a remote environment

Hm, how is that possible? You still need `drush` locally so you can do things like `drush @site updatedb`, which means you need to run composer install.

> Projects using composer scripts that still do things on install/update regardless of the state of dependencies

The example above only affects `composer install`, not `composer update`. If `composer install` has unpredictable output due to scripts, that feels like a bug in those scripts.

I actually set this up starting with the [task example that specifically uses composer](https://taskfile.dev/usage/#using-programmatic-checks-to-indicate-a-task-is-up-to-date). If that example is not OK, and we can find composer docs to support it, then perhaps we should go the other way and file a PR to Task with a different example for programmatic checks.

> Trying to debug problems in upstream dependencies

Wouldn't most developers be running `update` and not install?

Let's discuss! Hopefully before the Thursday call, but if not let's not let it slip past then.

**Comment by deviantintegral** [comment]
I'm coming back at this again for a new project.

Re-reading this issue and the code we have, I think this bit is the most important:

> As well, by consolidating to a single command for composer install for both prod and dev installs, it allows parent tasks to depend on a single task and set the environment variable if needed.

As a real world example: A developer (or CI!) may want to call `task drupal:update` or `task sync` after pulling code from git. However, they may have missed running `task composer:production` first, because in most cases dependencies aren't changing. Things break, and it ends up being a request for help from the team.

At an even higher level, what this leads to is an architecture decision for tasks: **that all tasks should express their dependencies to run as `deps: ` lines or by specifically calling `- task: <something>` in their commands.

Thoughts @justafish @mrdavidburns ?

**Comment by mrdavidburns** [comment]
If I'm understanding this correctly. I can't think of a situation where we would want to run `task drupal:update` without first running `composer install` or `composer install --dev`. I do feel there is a situation where we want one or the other (like dev on our locals). I would expect `task drupal:update` to at least run `composer install`. 

Could/Would we also have a separate task command like `task drupal-dev:update` or `task drupal:update:dev` that we can run locally which would do `composer install --dev`?

**Comment by deviantintegral** [comment]
> Could/Would we also have a separate task command like task drupal-dev:update or task drupal:update:dev that we can run locally which would do composer install --dev?

That's somewhat similar to what drainpipe has now, with `composer:development` and `composer:production` tasks. I think the difficulty is that it means you have to do this for _every_ task that you want to support both types with. composer, drupal, automated tests, custom tasks calling drush and generating reports... it adds up! Whereas, the use of an environment variable (and task variables) lets us pass settings through to an underlying task without duplication.
